sourcegraph/src-cli
# Add proxy support

This PR adds support for explicit HTTP(S), SOCKS5 and UNIX Domain Socket proxies that can be set using the `SRC_PROXY` environment variable:
```
SRC_ENDPOINT=https://sourcegraph.sourcegraph.com \
SRC_ACCESS_TOKEN=deadbeef \
SRC_PROXY=https://localhost:8080 \
src login --insecure-skip-verify=true

✔️  Authenticated as peter.guy on https://sourcegraph.sourcegraph.com
```

### Test plan

CI tests all pass.

Added more config tests.

#### Manual tests

*Option 1:*

Use `dev/test-proxies.sh`. 
Follow the instructions in comments at the beginning, then run the script.

*Option 2:*

0. Install `socat` and `mitmproxy`
```
brew install socat mitmproxy
```
3. In two separate terminals, run:
```
socat -d -d unix-listen:${HOME}/src-proxy.sock,fork tcp:localhost:8080
```
and
```
mitmproxy -v
```
4. Check out this branch and build `src-cli`:
```
go build -o ~/go/bin/src ./cmd/src
```
5. Run `src`, setting the endpoint, PAT, and UDS path in the environment variables:
```
SRC_ENDPOINT=https://sourcegraph.sourcegraph.com \
SRC_ACCESS_TOKEN=deadbeef \
SRC_PROXY=~/src-proxy.sock \
~/go/bin/src login --insecure-skip-verify=true
```
- You need to use the `--insecure-skip-verify=true` flag because `mitmproxy` uses a self-signed certificate. (You can get rid of that by [importing the `mitmproxy` CA cert into your local cert store](https://docs.mitmproxy.org/stable/concepts-certificates/))
- You should get a successful "Authenticated as ..." message. If the PAT is wrong, you'll see "Problem: Invalid access token."
- Look in the terminal where `socat` is running. You'll see output showing connections being made.
- Look in the terminal running `mitmproxy`. You'll see output showing connections being made.

Repeat those tests using the `mitmproxy` HTTP proxy directly:
```
SRC_ENDPOINT=https://sourcegraph.sourcegraph.com \
SRC_ACCESS_TOKEN=deadbeef \
SRC_PROXY=https://localhost:8080 \
~/go/bin/src login --insecure-skip-verify=true
```

To test SOCKS5 proxies, you can use a Docker-based SOCKS5 proxy.
Run this command in a terminal window:
```
docker run --rm -p 1080:1080 serjs/go-socks5-proxy
```
Then in another terminal window, run `src` pointing at that proxy:
```
SRC_ENDPOINT=https://sourcegraph.sourcegraph.com \
SRC_ACCESS_TOKEN=deadbeef \
SRC_PROXY=socks5://localhost:1080 \
~/go/bin/src login
```
You should see the same success messages as before, and in the terminal running the `docker` command, you should see connections being made.

We support proxy authentication with HTTP(S) and SOCKS5 proxies.

To test that, run `mitmproxy` and `serjs/go-socks-proxy` with authentication specified:
```
mitmproxy -v --proxyauth user:pass
```
```
docker run --rm -p 1080:1080 -e PROXY_USER=user -e PROXY_PASSWORD=pass serjs/go-socks5-proxy
```
Then add those credentials to the `SRC_PROXY` URL: either `https://user:pass@localhost:8080` or `socks5://user:pass@localhost:1080`

Also confirm that non-proxied connections can be made by omitting the `SRC_PROXY` environment variable setting.
